File: backend-refactor/src/routes/measurements.py
"""Measurement routes.

Endpoints
---------
- ``GET /measurements?signal_ids=1,2,3`` -- measurements for multiple signals
"""

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query

from models.measurement import FlatMeasurementList, MeasurementList, ResponseFormat
from services import get_measurement_service, get_signal_service
from services.measurement import MeasurementService
from services.signal import SignalService

router = APIRouter(prefix="/measurements", tags=["measurements"])


@router.get(
    "",
    response_model=MeasurementList | FlatMeasurementList,
    response_model_by_alias=False,
    responses={
        400: {"description": "Invalid signal IDs or date range"},
        404: {"description": "One or more signals not found"},
    },
)
async def get_measurements(
    signal_ids: str = Query(..., description="Comma-separated signal IDs"),
    from_date: datetime = Query(..., alias="from", description="Start datetime (ISO 8601)"),
    to_date: datetime = Query(..., alias="to", description="End datetime (ISO 8601)"),
    limit: int = Query(1000, ge=1, le=10000, description="Maximum results per page"),
    offset: int = Query(0, ge=0, description="Number of results to skip"),
    fmt: ResponseFormat = Query(
        ResponseFormat.OBJECTS, alias="format", description="Response layout: objects or flat"
    ),
    signal_svc: SignalService = Depends(get_signal_service),
    measurement_svc: MeasurementService = Depends(get_measurement_service),
) -> MeasurementList | FlatMeasurementList:
    """Return paginated measurements for one or more signals within a date range.

    The ``signal_ids`` query parameter accepts a comma-separated string of
    integer signal IDs (e.g. ``100,200,300``).

    Raises:
        HTTPException 400: If *signal_ids* is empty, contains non-integers,
            or the date range is invalid (``from >= to``).
        HTTPException 404: If any of the requested signal IDs are unknown.
    """
    stripped = [sid.strip() for sid in signal_ids.split(",") if sid.strip()]

    if not stripped:
        raise HTTPException(status_code=400, detail="At least one signal_id is required")

    MAX_SIGNAL_IDS = 100
    if len(stripped) > MAX_SIGNAL_IDS:
        raise HTTPException(
            status_code=400,
            detail=f"Too many signal_ids (max {MAX_SIGNAL_IDS})",
        )

    try:
        id_list = [int(sid) for sid in stripped]
    except ValueError:
        raise HTTPException(status_code=400, detail="All signal_ids must be integers") from None

    unknown = signal_svc.find_unknown_ids(id_list)
    if unknown:
        raise HTTPException(
            status_code=404,
            detail=f"Signal(s) {', '.join(map(str, unknown))} not found",
        )

    try:
        measurement_svc.validate_date_range(from_date, to_date)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from None

    return measurement_svc.get_measurements(
        id_list, from_date, to_date, limit=limit, offset=offset, fmt=fmt
    )

File: backend-refactor/alt/src/routes/measurements.rs
use std::sync::Arc;

use axum::extract::{Query, State};
use axum::Json;
use serde::Deserialize;

use crate::errors::AppError;
use crate::routes::signals::{parse_datetime, validate_limit, validate_offset};
use crate::service::AssetService;

const MAX_SIGNAL_IDS: usize = 100;

/// Query parameters for multi-signal measurements endpoint.
#[derive(Debug, Deserialize)]
pub struct MultiMeasurementsQuery {
    pub signal_ids: Option<String>,
    pub from: Option<String>,
    pub to: Option<String>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
    pub format: Option<String>,
}

/// GET /measurements
#[utoipa::path(
    get,
    path = "/measurements",
    params(
        ("signal_ids" = String, Query, description = "Comma-separated signal IDs"),
        ("from" = String, Query, description = "Start datetime (ISO 8601)"),
        ("to" = String, Query, description = "End datetime (ISO 8601)"),
        ("limit" = Option<i64>, Query, description = "Page size (1-10000, default 1000)"),
        ("offset" = Option<i64>, Query, description = "Offset (default 0)"),
        ("format" = Option<String>, Query, description = "Response format: objects or flat")
    ),
    responses(
        (status = 200, description = "Paginated measurements"),
        (status = 400, description = "Invalid parameters"),
        (status = 404, description = "Signal not found"),
        (status = 422, description = "Missing parameters")
    )
)]
pub async fn get_measurements(
    State(service): State<Arc<AssetService>>,
    Query(params): Query<MultiMeasurementsQuery>,
) -> Result<Json<serde_json::Value>, AppError> {
    // Validate signal_ids presence
    let signal_ids_str = params
        .signal_ids
        .ok_or_else(|| AppError::UnprocessableEntity("Missing required parameter: 'signal_ids'".to_string()))?;

    // Parse signal_ids with the specific validation order from spec:
    // 1. Empty check (400)
    // 2. Max count check (400)
    // 3. Integer parse validation (400)
    // 4. Existence checks (404) - done in service
    // 5. Date range validation (400) - done in service

    let segments: Vec<&str> = signal_ids_str
        .split(',')
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .collect();

    if segments.is_empty() {
        return Err(AppError::BadRequest(
            "At least one signal ID is required".to_string(),
        ));
    }

    if segments.len() > MAX_SIGNAL_IDS {
        return Err(AppError::BadRequest(format!(
            "Too many signal_ids (max {})",
            MAX_SIGNAL_IDS
        )));
    }

    let parsed_ids: Result<Vec<i64>, _> = segments.iter().map(|s| s.parse::<i64>()).collect();
    let signal_ids = parsed_ids.map_err(|_| {
        AppError::BadRequest("signal_ids must be valid integers".to_string())
    })?;

    // Validate from/to presence
    let from_str = params
        .from
        .ok_or_else(|| AppError::UnprocessableEntity("Missing required parameter: 'from'".to_string()))?;
    let to_str = params
        .to
        .ok_or_else(|| AppError::UnprocessableEntity("Missing required parameter: 'to'".to_string()))?;

    let format = params.format.unwrap_or_else(|| "objects".to_string());
    if format != "objects" && format != "flat" {
        return Err(AppError::UnprocessableEntity(format!(
            "Invalid format '{}'. Must be 'objects' or 'flat'",
            format
        )));
    }

    let limit = validate_limit(params.limit)?;
    let offset = validate_offset(params.offset)?;

    let from = parse_datetime(&from_str)?;
    let to = parse_datetime(&to_str)?;

    let result =
        service.get_multi_signal_measurements(&signal_ids, from, to, limit, offset, &format)?;
    Ok(Json(result))
}
